src/pycapsule.rs

```diff
@@ -0,0 +1,330 @@
+            Ok(unsafe { py.from_owned_ptr::<PyCapsule>(cap_ptr) })
+        }
+    }
+
+    /// Import an existing capsule.
+    ///
+    /// The `name` should match the path to module attribute exactly in the form
+    /// of `module.attribute`, which should be the same as the name within the
+    /// capsule. `no_block` indicates to use
+    /// [PyImport_ImportModuleNoBlock()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModuleNoBlock)
+    /// or [PyImport_ImportModule()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModule)
+    /// when accessing the capsule.
+    ///
+    /// ## Safety
+    /// This is unsafe, as there is no guarantee when casting `*mut void` into `T`.
+    pub unsafe fn import<'py, T>(py: Python<'py>, name: &CStr, no_block: bool) -> PyResult<&'py T> {
+        let ptr = ffi::PyCapsule_Import(name.as_ptr(), no_block as c_int);
+        if ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(&*(ptr as *const T))
+        }
+    }
+
+    /// Set a context pointer in the capsule to `T`
+    pub fn set_context<'py, T: 'static>(&self, py: Python<'py>, context: T) -> PyResult<()> {
+        let ctx = Box::new(context);
+        let result =
+            unsafe { ffi::PyCapsule_SetContext(self.as_ptr(), Box::into_raw(ctx) as _) as u8 };
+        if result != 0 {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Get a reference to the context `T` in the capsule, if any.
+    ///
+    /// ## Safety
+    ///
+    /// This is unsafe, as there is no guarantee when casting `*mut void` into `T`.
+    pub unsafe fn get_context<T>(&self, py: Python) -> PyResult<Option<&T>> {
+        let ctx = ffi::PyCapsule_GetContext(self.as_ptr());
+        if ctx.is_null() {
+            if self.is_valid() & PyErr::occurred(py) {
+                Err(PyErr::fetch(py))
+            } else {
+                Ok(None)
+            }
+        } else {
+            Ok(Some(&*(ctx as *const T)))
+        }
+    }
+
+    /// Obtain a reference to the value `T` of this capsule.
+    ///
+    /// # Safety
+    /// This is unsafe because there is no guarantee the pointer is `T`
+    pub unsafe fn reference<T>(&self) -> &T {
+        &*(self.get_pointer() as *const T)
+    }
+
+    /// Get the raw `c_void` pointer to the value in this capsule.
+    pub fn get_pointer(&self) -> *mut c_void {
+        unsafe { ffi::PyCapsule_GetPointer(self.0.as_ptr(), self.name().as_ptr()) }
+    }
+
+    /// Check if this is a valid capsule.
+    pub fn is_valid(&self) -> bool {
+        let r = unsafe { ffi::PyCapsule_IsValid(self.as_ptr(), self.name().as_ptr()) } as u8;
+        r != 0
+    }
+
+    /// Get the capsule destructor, if any.
+    pub fn get_destructor<T: 'static, D: FnOnce(T)>(&self) -> PyResult<Option<&D>> {
+        let val = unsafe { Box::from_raw(self.get_pointer() as *mut &CapsuleContents<T, D>) };
+        Ok(Some(&val.destructor))
+    }
```

Comment:
TBH this method looks nearly impossible to use correctly due to the type annotations. It's also `unsafe` because there's no guarantee the capsule holds these types. Perhaps better to just not have this, and let users go for raw FFI if they need this?

Reply:
Nice! I agree, it was struggling to understand how to get back the original function the user wrote (and would expect) from this method after wrapping it into `ffi::PyCapsule_Destructor` type.